## Content

## 01. Import libraries
## 02. Import data
## 03. 4.6 Exercise 

# 4.6 Merging and exporting data - Part 2

# 01. Import libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create path

path = r'C:\Users\Frederick\Documents\07-12-2023 Instacart Basket Analysis'

# 02. Import data

In [3]:
# Import data for orders_products

df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [4]:
# Q4.Checking total size/shape of orders_prior merged

df_ords_prods.shape

(32434489, 12)

## Q4. The shape of the imported dataframe is the same as the one I exported. Total rows = 32434489, total columns = 12.

In [5]:
# Import data for products

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'), index_col = False)

# 03. 4.6 Exercise

In [6]:
# Check the output for orders_products dataframe

df_ords_prods.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  user_order_number  \
0             0           0   2539329        1                  1   
1             0           0   2539329        1                  1   
2             0           0   2539329        1                  1   
3             0           0   2539329        1                  1   
4             0           0   2539329        1                  1   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   2                  8                     NaN       14084   
2                   2                  8                     NaN       12427   
3                   2                  8                     NaN       26088   
4                   2                  8                     NaN       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [7]:
# Check the output for products dataframe

df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [8]:
# Checking total size/shape of products dataframe

df_prods.shape

(49672, 6)

## Q5. The common id between orders & products dataframe and products dataframe is the product_id. I will be able to do an inner join to combine the data.

In [9]:
# Q5.Merging orders & prodcuts dataframe with products dataframe using inner join

df_merged_prods_in = df_ords_prods.merge(df_prods, on = 'product_id', indicator = 'exists')

In [10]:
# Check the output for orders_products merged with product ID dataframe

df_merged_prods_in.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  user_order_number  \
0             0             0   2539329        1                  1   
1             1             1   2398795        1                  2   
2             2             2    473747        1                  3   
3             3             3   2254736        1                  4   
4             4             4    431534        1                  5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   3                  7                    15.0         196   
2                   3                 12                    21.0         196   
3                   4                  7                    29.0         196   
4                   4                 15                    28.0         196   

   add_to_cart_order  reordered _merge  Unnamed: 0_y product_name  aisle_id  \
0                  1          0   both           195         Soda        77   
1                  1          1   both           195         Soda        77   
2                  1          1   both           195         Soda        77   
3                  1          1   both           195         Soda        77   
4                  1          1   both           195         Soda        77   

   department_id  prices exists  
0              7     9.0   both  
1              7     9.0   both  
2              7     9.0   both  
3              7     9.0   both  
4              7     9.0   both

In [11]:
df_merged_prods_in['exists'].value_counts()

exists
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [12]:
# Merging orders & products dataframe with products dataframe using outer join

df_merged_prods_out = df_ords_prods.merge(df_prods, on = 'product_id', how = 'outer')

In [13]:
# Check the output for orders_products with outer merged with product ID dataframe

df_merged_prods_out.head()

Unnamed: 0.1  Unnamed: 0_x   order_id  user_id  user_order_number  \
0           0.0           0.0  2539329.0      1.0                1.0   
1           1.0           1.0  2398795.0      1.0                2.0   
2           2.0           2.0   473747.0      1.0                3.0   
3           3.0           3.0  2254736.0      1.0                4.0   
4           4.0           4.0   431534.0      1.0                5.0   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                 2.0                8.0                     NaN         196   
1                 3.0                7.0                    15.0         196   
2                 3.0               12.0                    21.0         196   
3                 4.0                7.0                    29.0         196   
4                 4.0               15.0                    28.0         196   

   add_to_cart_order  reordered _merge  Unnamed: 0_y product_name  aisle_id  \
0                1.0        0.0   both         195.0         Soda      77.0   
1                1.0        1.0   both         195.0         Soda      77.0   
2                1.0        1.0   both         195.0         Soda      77.0   
3                1.0        1.0   both         195.0         Soda      77.0   
4                1.0        1.0   both         195.0         Soda      77.0   

   department_id  prices  
0            7.0     9.0  
1            7.0     9.0  
2            7.0     9.0  
3            7.0     9.0  
4            7.0     9.0

In [14]:
df_merged_prods_out['_merge'].value_counts()

_merge
both          32435059
left_only            0
right_only           0
Name: count, dtype: int64

## Q6. Using the merge flag, there are less count than in the original orders_products dataframe. Specifically, the merged (inner) data only has 32404859 records, where as the original orders_products has 32434489 records. That's 29630 records less. This means that there might be some product_id in orders_products dataframe but not in products dataframe, or vice versa. Or this could mean that there might be some missing data in product_id in the orders_products dataframe. 

In [16]:
# Exporting orders_products merged with ID as pickle file due to the size

df_merged_prods_in.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_productsID_combined.pkl'))

In [ ]:
[]